## Проектная работа по SQL

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Задача — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Первичное ознакомление с данными

In [2]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)
#создали функцию с целью удобства вывода данных

In [3]:
sql = '''SELECT * FROM books '''
select(sql)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
sql = '''SELECT * FROM authors '''
select(sql)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
sql = '''SELECT * FROM ratings '''
select(sql)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
sql = '''SELECT * FROM reviews '''
select(sql)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [7]:
sql = '''SELECT * FROM publishers '''
select(sql)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

## Задания

### 1. Посчитать, сколько книг вышло после 1 января 2000 года;

In [8]:
sql = '''SELECT COUNT(book_id)
         FROM books
         WHERE publication_date >= '2000-01-01' '''
select(sql)

count
0    821

Начиная с 1 января 2000 года опубликовалась 821 книга.

### Для каждой книги подсчёт количества обзоров и среднюю оценку;

In [9]:
sql = '''SELECT rev.book_id, COUNT(DISTINCT rev.review_id) AS count_rev, AVG(rat.rating) AS avg_rating
         FROM reviews AS rev
         JOIN ratings rat on rev.book_id = rat.book_id         
         group by rev.book_id
         
         '''
select(sql)

book_id  count_rev  avg_rating
0          1          2    3.666667
1          2          1    2.500000
2          3          3    4.666667
3          4          2    4.500000
4          5          4    4.000000
..       ...        ...         ...
989      996          3    3.666667
990      997          3    3.400000
991      998          4    3.200000
992      999          2    4.500000
993     1000          4    3.833333

[994 rows x 3 columns]

Для каждой книги посчитали количество обзоров и средний рейтинг.

### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (с той целью, чтобы исключить из анализа брошюры).

In [10]:
sql = '''SELECT pub.publisher, COUNT(DISTINCT b.book_id) AS count_book
         FROM publishers pub
         JOIN books b ON pub.publisher_id = b.publisher_id
         WHERE b.num_pages > 50
         GROUP BY pub.publisher
         ORDER BY count_book DESC
         LIMIT 1 '''
select(sql)


publisher  count_book
0  Penguin Books          42

В ходе работы с БД выяснили, что издательство "Penguin Books" выпустило наибольшее число книгу с толщиной более 50 страниц.

###  Определение автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.

In [13]:
sql = ''' select t.author, max(t.rat_avg) as rat_max from
(select s.author as author,avg(r.rating) as rat_avg, count(r.rating_id) as rat_cnt from
(select * from authors a
inner join books b on b.author_id = a.author_id) s
inner join ratings r on r.book_id = s.book_id
group by s.author) t
where t.rat_cnt > 50
group by t.author
order by rat_max desc
limit 1
'''
select(sql)

author   rat_max
0  J.K. Rowling/Mary GrandPré  4.288462

##  Подсчёт среднего количества обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
sql = '''SELECT AVG(s.review_count) FROM 
(SELECT r.username as username, COUNT(DISTINCT r.review_id) as review_count 
FROM reviews r
JOIN (SELECT t.username as username, COUNT(t.rating_id) as rating_count FROM ratings t
GROUP BY t.username
HAVING count(t.rating_id) > 50) s on s.username = r.username
GROUP BY r.username) s
'''
select(sql)

avg
0  24.333333

В среднем, 24 пользователя сделали обзор, при этом поставив более 50 оценок.

##  Вывод

1. Начиная с 1 января 2000 года опубликовалась 821 книга.
2. Для каждой книги посчитали количество обзоров и средний рейтинг.
3. Выяснили, что издательство "Penguin Books" выпустило наибольшее число книгу с толщиной более 50 страниц.
4. Diana Gabaldon является автором с самым высокой средней оценкой книг - в учет брались только те книги, которые имели 50+ оценок.
5. В среднем, 24 пользователя сделали обзор, при этом поставив более 50 оценок.